### PR TITLE
Added functionality required for unit testing in etch-tools

### PR DESCRIPTION
**Pull request by NelisDrost**
- save line numbers of parsed functions, for use in reporting on unit tests
- added 'get_functions' to EtchParser, for getting a list of functions with details
- correctly parse multi-type templates:
`a : Map<String, UInt32>;`
- correctly parse nested function calls, e.g.:  (fixes LDGR-414)
`a = b.get(c).get(d);`

### CODE REVIEW

File: fetchai/ledger/parser/etch.grammar [hunk 1]
```diff
@@ -18,41 +18,41 @@ instruction: "use" NAME [use_shard] ";"                         -> use_global
...
 
 parameter_block: "(" (typed_name ","?)* ")"
 
-typed_name: NAME ":" TYPE
+typed_name: NAME ":" type
 
 input_block: "(" (expression ",")* expression? ")"
 
-template_block: "<" TYPE ">"
+template_block: "<" type ">"
```
Comment (wkaluza) [inline fetchai/ledger/parser/etch.grammar]: this doesn't look right - shouldn't a template block be a comma-separated list of types?

File: fetchai/ledger/parser/etch.grammar [hunk 2]
```diff
@@ -18,41 +18,41 @@ instruction: "use" NAME [use_shard] ";"                         -> use_global
 
 instantiation: "var" NAME "=" expression ";"
```
Comment (wkaluza) [inline fetchai/ledger/parser/etch.grammar]: Not part of these changes, but `"var" typed_name "=" expression ";"` is also valid in Etch.

File: fetchai/ledger/parser/etch.grammar (hunk 1 above)
Comment (NelisDrost) [reply fetchai/ledger/parser/etch.grammar]: Correct - this isn't actually used though, it duplicates the correct definition below. Removed it.

File: fetchai/ledger/parser/etch.grammar (hunk 2 above)
Comment (NelisDrost) [reply fetchai/ledger/parser/etch.grammar]: Fixed + added test.

**Pull request merged**